example/src/main/scala/cilib/example/CSVExample.scala

```diff
@@ -0,0 +1,147 @@
+import cilib.pso._
+import eu.timepit.refined.auto._
+import scalaz.Scalaz._
+import scalaz._
+import scalaz.concurrent.Task
+import scalaz.effect.IO.putStrLn
+import scalaz.effect._
+import scalaz.stream._
+import spire.implicits._
+import spire.math.Interval
+
+object CSVExample extends SafeApp {
+
+  // An example showing how to compare multiple algorithms across multiple
+  // benchmarks and save the results to a CSV file.
+
+  val bounds = Interval(-5.12, 5.12) ^ 30
+  val rng = RNG.init(12L)
+
+  // Define the benchmarks
+  val absolute = Environment(
+    cmp = Comparison.dominance(Max),
+    eval = Eval.unconstrained(Benchmarks.absoluteValue[NonEmptyList, Double])
+  )
+
+  val ackley = Environment(
+    cmp = Comparison.dominance(Max),
+    eval = Eval.unconstrained(Benchmarks.ackley[NonEmptyList, Double])
+  )
+  val quadric = Environment(
+    cmp = Comparison.dominance(Max),
+    eval = Eval.unconstrained(Benchmarks.quadric[NonEmptyList, Double])
+  )
+
+  val spherical = Environment(
+    cmp = Comparison.dominance(Max),
+    eval = Eval.unconstrained(Benchmarks.spherical[NonEmptyList, Double])
+  )
+
+  // Define the problem streams
+  val absoluteStream = Runner.staticProblem("absolute", absolute.eval)
+  val ackleyStream = Runner.staticProblem("ackley", ackley.eval)
+  val sphericalStream = Runner.staticProblem("spherical", spherical.eval)
+  val quadricStream = Runner.staticProblem("quadric", quadric.eval)
+
+  // Define the guides for our PSO algorithms
+  val cognitive = Guide.pbest[Mem[Double], Double]
+  val gbestGuide = Guide.gbest[Mem[Double]]
+  val lbestGuide = Guide.lbest[Mem[Double]](3)
+
+  // Define our algorithms
+  val gbestPSO = gbest(0.729844, 1.496180, 1.496180, cognitive, gbestGuide)
+  val lbestPSO = gbest(0.729844, 1.496180, 1.496180, cognitive, lbestGuide)
+
+  // Define iterators for the algorithms
+  val gbestIter = Iteration.sync(gbestPSO)
+  val lbestIter = Iteration.sync(lbestPSO)
+
+  // Define the initial swarm
+  val swarm =
+    Position.createCollection(PSO.createParticle(x => Entity(Mem(x, x.zeroed), x)))(bounds, 20)
+
+  // A class to hold the results that we want to save at the end of each iteration
+  final case class Results(min: Double, average: Double)
+
+  // A performance measure that we apply to the collection at the end of each iteration.
+  def performanceMeasure =
+    measure[NonEmptyList[Entity[Mem[Double], Double]], Unit, Results](collection => {
+
+      val fitnessValues = collection
+        .map(x =>
+          x.pos match {
+            case Point(_, _) => Double.PositiveInfinity
+            case Solution(_, _, o) =>
+              o match {
+                case Single(f, _) =>
+                  f match {
+                    case Feasible(v)      => v
+                    case Infeasible(_, _) => Double.PositiveInfinity
+                    case Adjusted(_, _)   => Double.PositiveInfinity
+                  }
+                case Multi(_) => Double.PositiveInfinity
```

Comment:
I really want these cases hidden from the user in order to prevent people using the holes that scala allows to change this data. The `Position` is a very important structure for CIlib, so I would actually suggest that we allow this access only via a lens. The crux at the end of the day would be that the if the fitness is a feasible value, we get that otherwise we return the `PositiveInfinity` value.

This can be done with:

`Lenses._singleFitness[Double].composePrism(Lenses._feasible)`

which results in an `Optional[Position[Double], Double]`. The optic can then be used (via `getOption`) to get the value within an `Option` which we can then `getOrElse(Double.PositiveInfinity)` on.

```scala
val feasibleOptic = Lenses._singleFitness[Double].composePrism(Lenses._feasible)
val fitnessValues = collection.map(x => feasibleOptic.getOption(x).getOrElse(Double.PositiveInfinity))
```